File: informatika ismeretek/kozep/2019_may/c#/Ultrabalaton.cs
using System;
using System.IO;
using System.Collections.Generic;

namespace Erettsegi {
    public class Ultrabalaton {

        public static void Main(string[] args) {
            var lines = File.ReadAllLines("ub2017egyeni.txt");
            var versenyzok = new List<Versenyzo>();
        
            for(var i = 1; i < lines.Length; ++i) {
                versenyzok.Add(new Versenyzo(lines[i]));
            }
        
            Console.WriteLine("3. Feladat: Egyéni indulók: " + versenyzok.Count);
        
            var celbaertNoiVersenyzok = 0;
            foreach(var versenyzo in versenyzok) {
                if(versenyzo.befejezesSzazalek == 100 && versenyzo.kategoria == "Noi") {
                    ++celbaertNoiVersenyzok;
                }
            }
        
            Console.WriteLine("4. Feladat: Célbaért női versenyzők: " + celbaertNoiVersenyzok);
            Console.WriteLine("5. Feladat: Írd be egy versenyző nevét!");
        
            var bekertNev = Console.ReadLine();
            var indultEBekert = false;
            var bekertSzazaleka = -1;
            
            foreach(var versenyzo in versenyzok) {
                if(versenyzo.nev == bekertNev) {
                    indultEBekert = true;
                    bekertSzazaleka = versenyzo.befejezesSzazalek;
                    break;
                }
            }
            
            if(indultEBekert) {
                Console.WriteLine("Indult egyéniben? Igen");
                Console.WriteLine("Teljesítette a távot? " + (bekertSzazaleka == 100 ? "Igen" : "Nem"));
            }else {
                Console.WriteLine("Indult egyéniben? Nem");
            }
        
            var ferfiAtlagIdo = 0D;
            var ferfiakSzama = 0;
        
            foreach(var versenyzo in versenyzok) {
                if(versenyzo.befejezesSzazalek == 100 && versenyzo.kategoria == "Ferfi") {
                    ++ferfiakSzama;
                    ferfiAtlagIdo += versenyzo.idoOraban();
                }
            }
        
            Console.WriteLine("7. Feladat: Átlagos idő: " + (ferfiAtlagIdo / ferfiakSzama));
        
            var ferfiGyoztes = versenyzok[0];
            var noiGyoztes = versenyzok[0];
        
            foreach(var versenyzo in versenyzok) {
                if(versenyzo.befejezesSzazalek == 100) {
                    if(versenyzo.kategoria == "Noi") {
                        if(versenyzo.idoOraban() < noiGyoztes.idoOraban()) {
                            noiGyoztes = versenyzo;
                        }
                    }else {
                        if(versenyzo.idoOraban() < ferfiGyoztes.idoOraban()) {
                            ferfiGyoztes = versenyzo;
                        }
                    }
                }
            }
        
            Console.WriteLine($"Nők: {noiGyoztes.nev} ({noiGyoztes.rajtszam}) - {noiGyoztes.ido}");
            Console.WriteLine($"Férfiak: {noiGyoztes.nev} ({noiGyoztes.rajtszam}) - {noiGyoztes.ido}");
        }
    }

    public class Versenyzo {
        public readonly string nev;
        public readonly int rajtszam;
        public readonly string kategoria;
        public readonly string ido;
        public readonly int befejezesSzazalek;
        
        public Versenyzo(string line) {
            var split = line.Split(';');
            
            nev = split[0];
            rajtszam = int.Parse(split[1]);
            kategoria = split[2];
            ido = split[3];
            befejezesSzazalek = int.Parse(split[4]);
        }
        
        public double idoOraban() {
            var daraboltIdo = ido.Split(':');
            var ora = double.Parse(daraboltIdo[0]);
            var perc = double.Parse(daraboltIdo[1]);
            var mp = double.Parse(daraboltIdo[2]);
            
            return ora + (perc / 60D) + (mp / 3600D);
        }
    }
}

File: informatika_ismeretek/kozep/2019_may/python/Ultrabalaton.py
from statistics import mean

class Versenyzo:
    def __init__(self, line):
        split = line.split(";")

        self.nev = split[0]
        self.rajtszam = int(split[1])
        self.kategoria = split[2]
        self.ido = split[3]
        self.befejezesSzazalek = int(split[4])

    def idoOraban(self):
        daraboltIdo = self.ido.split(":")
        ora = float(daraboltIdo[0])
        perc = float(daraboltIdo[1])
        mp = float(daraboltIdo[2])

        return ora + (perc / 60.0) + (mp / 3600.0)

with open("ub2017egyeni.txt", "r") as file:
    lines = file.readlines()
    versenyzok = [ Versenyzo(lines[i]) for i in range(1, len(lines)) ]

print(f"3. Feladat: Egyéni indulók: {len(versenyzok)}")

celbaertNoiVersenyzokSzama = sum(1 for k in versenyzok if k.befejezesSzazalek == 100 and k.kategoria == "Noi")
print(f"4. Feladat: Célbaért női versenyzők: {celbaertNoiVersenyzokSzama}")

bekertNev = input("5. Feladat: Írd be egy versenyző nevét! ")
bekertNevuVersenyzo = next((versenyzo for versenyzo in versenyzok if versenyzo.nev == bekertNev), None)

if bekertNevuVersenyzo is not None:
    print("    Indult egyéniben? Igen")
    teljesitette = "Igen" if bekertNevuVersenyzo.befejezesSzazalek == 100 else "Nem"
    print(f"    Teljesítette a távot? {teljesitette}")
else:
    print("    Indult egyéniben? Nem")

ferfiAtlagIdo = mean(k.idoOraban() for k in versenyzok if k.befejezesSzazalek == 100 and k.kategoria == "Ferfi")
print(f"7. Feladat: Átlagos idő: {ferfiAtlagIdo} óra")

noiGyoztes =   min((k for k in versenyzok if k.befejezesSzazalek == 100 and k.kategoria == "Noi"),   key = lambda k: k.idoOraban())
ferfiGyoztes = min((k for k in versenyzok if k.befejezesSzazalek == 100 and k.kategoria == "Ferfi"), key = lambda k: k.idoOraban())

print(f"Nők: {noiGyoztes.nev} ({noiGyoztes.rajtszam}) - {noiGyoztes.ido}")
print(f"Férfiak: {ferfiGyoztes.nev} ({ferfiGyoztes.rajtszam}) - {ferfiGyoztes.ido}")